evaluation.py

```diff
@@ -67,26 +68,32 @@ def batch_predict(i, x):
                     x[:, i:i+eval_batch_size, ...],
                     model.decoder.embedding.weight,
                     return_indexes=True,
+                    top_k=top_k_val,
                 )
             _preds = [
                 batch_predict(i, pred_embeds)
                 for i in range(0, pred_embeds.shape[1], eval_batch_size)]
             preds = torch.cat(_preds, dim=1)
-            correct += (preds.view(-1) == batch.trg[1:,:].view(-1)).sum().item()
         else:
-            preds = scores[:-1,:,:].argmax(2).squeeze()
-            correct += (preds.view(-1) == batch.trg[1:,:].view(-1)).sum().item()
+            preds = torch.topk(scores[:-1,:,:], top_k_val, sorted=True).indices
+        correct += [
```

Comment:
I'm a fan of one-liners, but this is too much. It's hard to read and without playing this code in  a terminal it is hard to understand how (or if) this works properly.

Either break this down into steps in a for loop or use vectorial operations exclusively.

Reply:
So would this be preferred:
```pyhon
for k in range(1, top_k_val + 1):
  correct[k-1] = (preds[..., :k] == batch.trg[1:, :, : np.newaxis]).any(-1).sum().item()
```

Reply:
No, each step of the list comprehension should be assigned to a variable with a name that describes what it is with respect to the process. For example:

????? = preds[..., :k] == batch.trg[1:,:,: np,newaxis]
xxxxx = ????.any(-1)
yyyyy = xxxxx.sum().item()

Where ?????, xxxxxx, yyyy explain what the result of this step is. The code should read like your explaining the steps to me out loud

Reply:
Too verbose?
```
for k in range(1, top_k_val + 1):
  correct_tokens = (preds[..., :k] == batch.trg[1:, :, : np.newaxis])
  correct_tokens_in_top_k = correct_tokens.any(-1)
  total_correct_tokens = correct_tokens_in_top_k.sum().item()
  correct[k-1] = total_correct_tokens
```

Reply:
Specifically I don't understand why you're taking the first K vectors of the predictions in an iterative fashion (keeping more and more of them)

Reply:
I am calculating top-k accuracy. The last dimension of `preds` holds the top `k` predictions (sorted). So if I index `preds` with `[:3]`, it will get me the model's top 3 predictions. If I then reduce with `.any(-1)`, it will tell me if any of the top 3 predictions were correct. I have to do this iteratively because, as far as I know, it is not possible to vector slice in PyTorch/NumPy (e.g., `preds[:, [:1, :2, :3]]` to get the top 1, 2, and 3).

Reply:
Ok, that's perfect!